### DISCUSSION TITLE
JetStream durable consumer does not receive messages when reconnecting after offline period - comparison with STAN durable queues. Is there a way to achieve similar behaviour?

### DISCUSSION THREAD

**Discussion by SzymonSzM** [nats-io/nats.net.v1#758] [Q&A]
Hello,

We have a problem to achieve with JetStream same behavior as with STAN connected to durable, queued subscription. Please let me write a scenario below:

1. Create a publisher and subscriber.
2. Publish message.
3. Subscriber receives message.
4. Crash subscriber.
5. Publish 3 messages.
6. Turn on subscriber.
7. Stan subscriber in this moment receives all 3 messages. JetStream subscriber in same scenario receives only last one.

More details:
 - implementation done in c#
 - STAN subscriptions done with durable name and queue name as we have microservices with many replicas
 - subscription in JetStream done as shown below, but it does not work: while no subscribers are offline, when comming back online messages are never resent - except for the last one

```csharp
cc = ConsumerConfiguration.Builder()
        .WithFilterSubject("subjectName")
        .WithDeliverSubject("deliverSubjectName")
        .WithDurable("durableName")
        .WithDeliverGroup("deliverGroup")
        .WithAckWait(1000)
        .WithAckPolicy(AckPolicy.Explicit)
        .WithDeliverPolicy(DeliverPolicy.All) //here was tested with All, Last, New settings
        .Build();
        
JetStreamManagementContext.AddOrUpdateConsumer("streamName", cc);

 PushSubscribeOptions pso = PushSubscribeOptions.BindTo("streamName", 
    "durableName");
                
IJetStreamPushAsyncSubscription sub = JetStreamManagementContext.PushSubscribeAsync(
    "subjectName",
    "deliverGroup",
    async (obj, args) => { await HandleReceivedMessage(args, onError); },
    false, 
    pso );

    sub.Start();

'''

Can anyone help and explain if there is a way to achieve simillar behaviour on JetStream like on STAN?

**Comment by scottf** [comment]
@SzymonSzM Transferred this discussion to the nats.net repo

**Comment by scottf** [comment]
From you example I can't tell exactly when your client was crashed and if you had more than once instance of the consumer running, but I'll try my best.

A couple things going on. You have a durable consumer, which means that once the client acks a message it will never see it again. The other thing you have turned on is auto ack in the subscribe, which means that if you don't ack the message in the handler, as soon as your handler code is finished, the client will ack the message for you.

Since you are using a deliver group, all subscriptions using this consumer get the messages round robin from the server. If all things are running normally, and for instance you have 3 clients running the same consumer, each client will only receive 1/3 of the messages.
If a client fails to ack a given message within the ack wait period, then it will be set for redelivery and will go to any one of the 3 clients the next chance.

Something to try while your subscriber is crashed is to run the NATS Cli and interrogate the consumer.
```
nats c info streamName durableName
-or-
nats consumer info streamName durableName
```
and you will be able to see the delivery status of messages.

**Reply by SzymonSzM** [reply]
Hi Scot, thank you for this answer. I checked it with CLI and found out that there was always pending one message even if i send many of them. Then i found out that stream was configured with - .WithMaxMessages(1) which i think made the subscription could receive only one message when after changing value to -1 everything start to work fine. Thank you very much for you answer as it lead me to fast fix.
